### DISCUSSION TITLE
Custom maps

### DISCUSSION THREAD

**Discussion by sultim-t** [sultim-t/xash-rt#150] [Announcements]
*Note: this is intended to tweak maps; if modifying becomes too complex for a mod, it might be worth to wait for Half-Life RTX (another project that adds path tracing to HL1: a full mod support is claimed), or NVIDIA RTX Remix*

*Note: [Texture replacements](https://github.com/sultim-t/xash-rt/discussions/124) is expected to be read first*

In HL1, each map contains static (walls, floors etc) and dynamic objects (characters. particles etc).
In HL1RT, static objects and static light sources can be exported to [gITF2 format](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#motivation) via RTGL1 devmode (more about devmode [here](https://github.com/sultim-t/xash-rt/discussions/124)).
With gITF, we can modify existing light sources, geometry; and even add new ones.

The overall workflow for a new map:
1. Export a map to gITF
1. Modify the glTF file (e.g. in Blender)
1. When the game renders a map, the corresponding gITF2 file is loaded and being rendered instead of the original map
1. Optional per-map properties can be tweaked in `data/scenes,json`

## Details
For example, we would like to make [this map](https://gamebanana.com/mods/59556) RT-compatible
![image](https://user-images.githubusercontent.com/31703959/224694846-47c475fc-4c64-4c43-a916-472fe1288c17.png)

Install it, by unzipping it to `Half-Life/valve` folder, so `dm_dust2_hl1.bsp` is inside `Half-Life/valve/maps`. Launch the game, load into it.

### Export
Ensure that `No server cull` is enabled (otherwise not all lights/objects will be exported).
![image](https://user-images.githubusercontent.com/31703959/224696083-086d71ec-9b30-422a-9469-7a6039827819.png)

In devmode window, click `Export frame geometry`
![image](https://user-images.githubusercontent.com/31703959/224696252-83b0eb31-cbfd-4eb6-8ffd-d11f03d954e1.png)

### Open
Open the exported glTF in Blender
![image](https://user-images.githubusercontent.com/31703959/224696815-2d65e82d-11a8-47c3-94ca-57a80e4d6e8d.png)

Tip: Enable viewport shading mode (to see the lighting/textures), and set sampling to 1 (so it's less laggy)
![image](https://user-images.githubusercontent.com/31703959/224698008-fdb4b750-1c99-4035-9c5c-daa18da81165.png)

Now we should see the map
![image](https://user-images.githubusercontent.com/31703959/224696958-a0dff34a-2e26-4869-af04-ce0d1a3dc008.png)

### Modify
All the objects must be inside `rtgl1_main_root` (as HL1 uses non-metric system)
![image](https://user-images.githubusercontent.com/31703959/224698270-5892c033-c172-4b20-be96-390335337ebc.png)

Let's make sun green
![image](https://user-images.githubusercontent.com/31703959/224698503-98c92b84-f7c5-4ef8-b4c8-cec4705aff2b.png)

### Export
HL1RT hot-reloads glTF files, so a game can just run in background, and as soon we export glTF, we should see the changes in the game instantly (if there are no errors, and if *devmode is enabled*). 

Click *File->Export->glTF 2.0* (it's useful to keybind it, e.g. *Ctrl-E*)
![image](https://user-images.githubusercontent.com/31703959/224699119-7fe6721f-c387-4eeb-a640-8d5f3525fb96.png)

Find the existing file, select it
![image](https://user-images.githubusercontent.com/31703959/224700570-d58035b4-9e32-4648-afc6-3a2b5599ada5.png)

glTF export settings must be:
1. Format = *glTF Separate*
1. Keep original = *True*
1. Include - Custom Properties = *True*
1. Include - Punctual Lights = *True*
1. Mesh - Tangents = *True*
    
![image](https://user-images.githubusercontent.com/31703959/224701379-d176c1d8-3c2a-4de4-9d87-a91beefb059d.png)

Tip: it's useful to save such export settings:
![image](https://user-images.githubusercontent.com/31703959/224702388-5611cb39-eb6a-4d7f-b633-0826997f2bdc.png)

The name line will be red, as we replace existing file. Click *Export*
![image](https://user-images.githubusercontent.com/31703959/224701756-94e55061-cadd-4c0b-8f6b-1876f506ffe0.png)

We would see a message in HL1 console
![image](https://user-images.githubusercontent.com/31703959/224702030-37e9c6f7-b820-4c36-a354-e0aace0553db.png)

and the sun is green now
![image](https://user-images.githubusercontent.com/31703959/224702159-78f54173-5e61-42a7-bfae-aaea63b5d30d.png)

### Cleaning
HLT1RT exporter juts works.

glTF doesn't support polygon lights, so they are approximated by spotlights, and usually the HL1RT exporter produces too many unnecessary lights around polygons. They should be removed (less lights = less noise)

Example: remove three unnecessary spotlights and leave only one point light
![image](https://user-images.githubusercontent.com/31703959/224703824-6bc842f0-c733-4960-9563-445ffc4088d0.png)

Just works
![image](https://user-images.githubusercontent.com/31703959/224704447-6cc51dc4-9e09-423b-a8bc-172765faa018.png)

### Per-scene properties
There's a limited set of optional properties that can be set for a scene: `Half-Life/rt/data/scenes.json` contains the properties. If a property is not found in `scenes.json`, then values from cvars are used.

For custom maps, add an entry with a specified `sceneName`.

![image](https://user-images.githubusercontent.com/31703959/224705643-e4194cbc-b71f-4ccb-8450-a1fe9a8d4648.png)

Note: HL1 map names have format `maps/c0a0.bsp`, but `sceneName` is only `c0a0`
Warning: there are bugs related to spaces in the JSON-parser, specifically, spaces around `:` in arrays are interpreted as errors. Remove spaces and look into the console logs for parser errors!

### Per-object properties
glTF allows extra fields on objects. We can utilize that to provide more info to the renderer. For example, make a light volumetric, or mark a mesh as mirror.
![image](https://user-images.githubusercontent.com/31703959/224714508-7b5dfba0-b068-4246-9048-d37626feeff8.png)

Lights:
* `lightstyle` -- use dynamic intensity mu;tiplier from original HL1 map lightstyles
* `isVolumetric` -- only one light can be marked as volumetric per map; if none of the glTF lights have this property, sun is used
* todo

Meshes:
* `isGlass`
* `isMirror`
* `isWater`
* todo

### Adding custom objects
To add a new mesh, just create it in Blender and put it inside `rtgl1_main_root`.
Custom textures (TGA/PNG) should exist in the same folder as its `.gltf` file.
**Never** use textures from `mat_junction` folder for new objects.

### Light leaks
When using volumtrics, light might leak throw thin walls or if nothing obstructs the light source (e.g. a building doesn't have a roof). To prevent such leaks, walls should be thicker (or rt_volume_far smaller, so the grid voxels are smaller) / add geometry to obstruct the light source.

### Publish
Only distribute `scenes.json`, `.gltf`, `.bin` and your own textures.
**Never** include `mat_junction` folder as it contains original HL1 textures, and used *only* to display textures in Blender.

### Screenshots
Editing
![image](https://user-images.githubusercontent.com/31703959/224709656-6c0776ff-7bc4-4a61-8ae5-f3855dba6b2e.png)

Different light setups
![Untitled](https://user-images.githubusercontent.com/31703959/224710291-8dd87baf-b1b3-4aa0-98b8-8de74c7ef294.jpg)

**Comment by THG327** [comment]
I converted the Hazard-Course and it was surprisingly easy to do following this guide.
[rt-training.zip](https://github.com/sultim-t/xash-rt/files/10983441/rt-training.zip)
I also noticed that you can have individual textures.json files for scenes. Would it be possible to have the per-scene properties loaded from those individual folders instead of overwriting the main scenes.json file?
